queue/driver_mongo.go

```diff
@@ -1400,14 +1268,146 @@ type dispatchAttemptInfo struct {
 	// misses is the number of times the dispatcher attempted to dispatch a job
 	// but failed.
 	misses int
+	// total is the total number of documents the cursor returned.
+	total int
+}
+
+// dispatchStaleAndPending attempts to dispatch a job. Stale in progress jobs are given precedence over pending jobs.
+func (d *mongoDriver) dispatchStaleAndPending(ctx context.Context, startAt time.Time) (amboy.Job, dispatchAttemptInfo, error) {
+	now := time.Now()
+	var info dispatchAttemptInfo
+
+	job, dispatchInfo, err := d.tryDispatchJob(ctx, d.getNextStaleInProgressQuery(now), d.opts.SampleSize, startAt)
+	info.misses += dispatchInfo.misses
+	info.skips += dispatchInfo.skips
+	if err != nil {
+		return nil, info, errors.Wrap(err, "dispatching stale in progress job")
+	}
+
+	if job != nil || d.opts.SampleSize > 0 && dispatchInfo.total >= d.opts.SampleSize {
+		return job, info, nil
+	}
+
+	pendingSampleSize := 0
+	if d.opts.SampleSize > 0 {
+		pendingSampleSize = d.opts.SampleSize - dispatchInfo.total
+	}
+	job, dispatchInfo, err = d.tryDispatchJob(ctx, d.getNextPendingQuery(now), pendingSampleSize, startAt)
+	info.misses += dispatchInfo.misses
+	info.skips += dispatchInfo.skips
+	if err != nil {
+		if err != nil {
+			return nil, info, errors.Wrap(err, "dispatching stale in progress job")
+		}
+	}
+
+	return job, info, nil
+}
+
+// getNextStaleInProgressQuery returns the query for the next available stale in-progress jobs.
+func (d *mongoDriver) getNextStaleInProgressQuery(now time.Time) bson.M {
+	staleBefore := now.Add(-d.LockTimeout())
+	qd := d.getInProgQuery(bson.M{"status.mod_ts": bson.M{"$lte": staleBefore}})
+	d.modifyQueryForGroup(qd)
```

Comment:
Just a suggestion that you could turn this into a nicer-looking builder if `modifyQueryForGroup` returns the `bson.M` query to do `d.modifyQueryForGroup(d.modifyQueryForTimeLimits(qd, now))`.